File: next-js-boilerplate/src/lib/currency.test.ts
import { describe, it, expect } from "vitest";
import { formatCurrency, formatPrice, toCurrencyCode } from "@/lib/currency";

describe("toCurrencyCode", () => {
  it("passes through known currency codes, case-insensitively", () => {
    expect(toCurrencyCode("USD")).toBe("USD");
    expect(toCurrencyCode("eur")).toBe("EUR");
    expect(toCurrencyCode("Try")).toBe("TRY");
  });

  it("falls back to USD for an unrecognized currency", () => {
    expect(toCurrencyCode("GBP")).toBe("USD");
    expect(toCurrencyCode("")).toBe("USD");
  });
});

describe("formatCurrency", () => {
  it("formats a one-time amount with no recurring suffix", () => {
    expect(formatCurrency(1999, "USD")).toBe("$19.99");
    expect(formatCurrency(0, "USD")).toBe("$0.00");
  });

  it("formats non-USD currencies in their own locale", () => {
    expect(formatCurrency(1899, "EUR")).toContain("18,99");
    expect(formatCurrency(34999, "TRY")).toContain("349,99");
  });
});

describe("formatPrice", () => {
  it("appends a /mo cadence for a non-zero recurring price", () => {
    expect(formatPrice(999, "USD")).toBe("$9.99/mo");
  });

  it("special-cases zero as Free, no cadence suffix", () => {
    expect(formatPrice(0, "USD")).toBe("Free");
  });
});

File: flutter-boilerplate/test/lib/currency_test.dart
import 'package:flutter_boilerplate/lib/currency.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('toCurrencyCode', () {
    test('uppercases a known currency', () {
      expect(toCurrencyCode('eur'), 'EUR');
    });

    test('falls back to USD for an unsupported currency', () {
      expect(toCurrencyCode('gbp'), 'USD');
    });
  });

  group('formatPrice', () {
    test('renders "Free" for zero cents, regardless of currency', () {
      expect(formatPrice(0, 'EUR'), 'Free');
    });

    test('formats USD with the /mo cadence suffix', () {
      expect(formatPrice(999, 'USD'), '\$9.99/mo');
    });

    test('formats TRY using the Turkish locale grouping', () {
      // tr_TR uses a comma decimal separator, unlike en_US.
      expect(formatPrice(1999, 'TRY'), contains('19,99'));
    });
  });

  group('formatCurrency', () {
    test('does not add the /mo suffix (unlike formatPrice)', () {
      expect(formatCurrency(500, 'USD'), '\$5.00');
    });
  });
}
